addons/sale_loyalty/models/sale_order.py

```diff
@@ -913,8 +913,16 @@ def _program_check_compute_points(self, programs):
                         points += rule.reward_point_amount
                     elif rule.reward_point_mode == 'money':
                         # Compute amount paid for rule
-                        # NOTE: this does not account for discounts -> 1 point per $ * (100$ - 30%) will result in 100 points
-                        amount_paid = sum(max(0, line.price_total) for line in order_lines if line.product_id in rule_products)
+                        # NOTE: this accounts for discounts -> 1 point per $ * (100$ - 30%) will result in 70 points
+                        lines_for_points = []
+                        for line in self.order_line - self._get_no_effect_on_threshold_lines():
+                            is_discount = line.reward_id.reward_type == 'discount'
+                            reward_program = line.reward_id.program_id
+                            if line.reward_id and reward_program.program_type in ['ewallet', 'gift_card', program.program_type]:
```

Comment:
Hope I can make it make sense :sweat_smile: 

Just to recall, at this point in the code we have the following:
- The rule of the program grants points https://github.com/odoo/odoo/blob/1a09bd96dba8f05794bd694b24344019ef50a6ef/addons/sale_loyalty/models/sale_order.py#L894-L895
- The program applies on either `current orders` or `current & future` and the points are not. We are in the else of this if condition. https://github.com/odoo/odoo/blob/1a09bd96dba8f05794bd694b24344019ef50a6ef/addons/sale_loyalty/models/sale_order.py#L897
- The program grans points per $ spent https://github.com/odoo/odoo/blob/1a09bd96dba8f05794bd694b24344019ef50a6ef/addons/sale_loyalty/models/sale_order.py#L914

Technically you can apply multiple coupons but not from the same program. I've tested the scenarios and I get a message like **"This promotion was already applied"**. Promotions and discount programs usually are set to grant 1 point per order if some conditions are met. So they do not really enter this part of the code.

This setting (what the program applies on + granting points per $ spent) is usually seen with the loyalty program type. This is also usually (functionally) the only program type that will have multiple different rewards.

> Would you mind explaining me why you're also skipping programs with the same type? 

Well this is mostly for the loyalty program. From the discussion I've had with DALA I understand that rewards that were obtained by spending loyalty points should not have the effect of diminishing the amount of points awarded. Here a more concrete example. Let's say you exchange 1000 loyalty points for a 10$ discount on your order. The order is simply 1 product of 50$. You should still get 50 * `nbr points per $`. 

I could write it like (with your last comment taken into account)
```python
if line.reward_id.program_id.program_type in ['ewallet', 'gift_card']:
    continue
if program.program_type == "loyalty" and line.reward_id.program_id.program_type == "loyalty"
```
This is more clearer to understand but maybe it should still include all programs. Let's say it's possible that through dev mode some configurations are so that there is a coupons program that would grant points per dollars spent if rule conditions are met. W just assume rule conditions are met. He gets 0.01 point per $ spent, he has an order of 100$, so he should get 1 point that he can use for the coupon (since coupons discount rewards are set in exchange of x points, usually 1). His order then is 90$ which means he should get 0.9 points and not 1 that he had spent.

This is a weird setting for discounts I can conceive but we've had customer that had loyalty programs that were like:
You get 1 point per $ spent and you can exchange 100 points for 1 $ on your orders. The program was set to applies on current and future orders. So for a 100$ order the clients should get 100 point that he can decide to use immediately but then his total is 99 and not 100 so he doesn't have the 100 point to spend immediately ( without the fix). This is the use case I discussed with DALA when he said that we should not deduct points twice (the use of points to get the reward and the deduction of points because of the reward).

Sorry if it's not clear. Feel free to message me for a call if you prefer :smile:

Reply:
Got it, thanks for the detailed explanation ! :)